force close on "save as" option

**Describe the bug**
the "save as" option among android share options cause the app to break.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to 'Firefox> Share'
2. Click on 'Save as...'
4. amaze file launches. as i select the floating button, it force closes.

**Expected behavior**
maybe amaze could've downloaded the content

**Smartphone (please complete the following information):**
 - Device: LG G4 H818P
 - OS: nougat
 - Rooted: no
 - Version: 3.4.3

**Additional context**
thanks for the app though.

I went to a random site and tried to save it using the Save as option, from the share button and I got the below error

`2020-07-26 08:59:02.373 18710-18745/com.amaze.filemanager.debug D/EGL_emulation: eglMakeCurrent: 0xdf2c7320: ver 3 0 (tinfo 0xea489a00)
2020-07-26 08:59:06.567 18710-18755/com.amaze.filemanager.debug E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
    Process: com.amaze.filemanager.debug, PID: 18710
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$4.done(AsyncTask.java:399)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.lang.NullPointerException: uri
        at com.android.internal.util.Preconditions.checkNotNull(Preconditions.java:133)
        at android.content.ContentResolver.openInputStream(ContentResolver.java:1170)
        at com.amaze.filemanager.filesystem.FileUtil$1.doInBackground(FileUtil.java:213)
        at com.amaze.filemanager.filesystem.FileUtil$1.doInBackground(FileUtil.java:202)
        at com.amaze.filemanager.application.AppConfig$1.doInBackground(AppConfig.java:135)
        at android.os.AsyncTask$3.call(AsyncTask.java:378)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
2020-07-26 08:59:08.155 18710-18745/com.amaze.filemanager.debug D/EGL_emulation: eglMakeCurrent: 0xdf2c7320: ver 3 0 (tinfo 0xea489a00)`

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -56,8 +56,7 @@
             android:label="@string/appbar_name"
             android:launchMode="singleInstance"
             android:name=".ui.activities.MainActivity"
-            android:theme="@android:style/Theme.Translucent.NoTitleBar"
-            android:screenOrientation="sensor">
+            android:theme="@android:style/Theme.Translucent.NoTitleBar">
 
             <intent-filter android:label="@string/appbar_name">
                 <action android:name="android.intent.action.MAIN" />

--- a/app/src/main/java/com/amaze/filemanager/application/AppConfig.java
+++ b/app/src/main/java/com/amaze/filemanager/application/AppConfig.java
@@ -87,11 +87,10 @@ public class AppConfig extends GlideApplication {
     utilsProvider = new UtilitiesProvider(this);
     utilsHandler = new UtilsHandler(this, utilitiesDatabase);
 
-    // FIXME: in unit tests when AppConfig is rapidly created/destroyed this call will cause
-    // IllegalThreadStateException.
-    // Until this gets fixed only one test case can be run in a time. - Raymond, 24/4/2018
     backgroundHandlerThread.start();
     backgroundHandler = new Handler(backgroundHandlerThread.getLooper());
+
+    runInBackground(jcifs.Config::registerSmbURLHandler);
 
     // disabling file exposure method check for api n+
     StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();

--- a/app/src/main/java/com/amaze/filemanager/database/UtilsHandler.java
+++ b/app/src/main/java/com/amaze/filemanager/database/UtilsHandler.java
@@ -58,6 +58,8 @@ import androidx.annotation.NonNull;
  */
 public class UtilsHandler {
 
+  private static final String TAG = UtilsHandler.class.getSimpleName();
+
   private final Context context;
 
   private final UtilitiesDatabase utilitiesDatabase;
@@ -313,6 +315,13 @@ public class UtilsHandler {
   }
 
   public void renameSMB(String oldName, String oldPath, String newName, String newPath) {
+    try {
+      oldPath = SmbUtil.getSmbEncryptedPath(AppConfig.getInstance(), oldPath);
+      newPath = SmbUtil.getSmbEncryptedPath(AppConfig.getInstance(), newPath);
+    } catch (GeneralSecurityException | IOException e) {
+      Log.e(TAG, "Error encrypting SMB path", e);
+    }
+
     SmbEntry smbEntry = utilitiesDatabase.smbEntryDao().findByNameAndPath(oldName, oldPath);
     smbEntry.name = newName;
     smbEntry.path = newPath;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
@@ -628,7 +628,6 @@ public abstract class FileUtil {
     try {
       f.createNewFile();
       out = new FileOutputStream(f, false);
-
       outputWriter = new OutputStreamWriter(out);
       outputWriter.write(data);
       return true;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
@@ -29,6 +29,7 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.io.OutputStreamWriter;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.MalformedURLException;
@@ -617,6 +618,35 @@ public abstract class FileUtil {
       }
     }
     return false;
+  }
+
+  public static boolean mktextfile(String data, String path, String fileName) {
+
+    File f = new File(path + "/" + fileName + ".txt");
+    FileOutputStream out = null;
+    OutputStreamWriter outputWriter = null;
+    if (f.exists()) {
+      f.delete();
+    }
+    try {
+      f.createNewFile();
+      out = new FileOutputStream(f);
+
+      outputWriter = new OutputStreamWriter(out);
+      outputWriter.write(data);
+      return true;
+    } catch (IOException io) {
+      Log.e(FileUtil.LOG, io.getMessage());
+      return false;
+    } finally {
+      try {
+        outputWriter.close();
+        out.flush();
+        out.close();
+      } catch (IOException e) {
+        Log.e(FileUtil.LOG, e.getMessage());
+      }
+    }
   }
 
   /**

--- a/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/FileUtil.java
@@ -625,12 +625,9 @@ public abstract class FileUtil {
     File f = new File(path + "/" + fileName + ".txt");
     FileOutputStream out = null;
     OutputStreamWriter outputWriter = null;
-    if (f.exists()) {
-      f.delete();
-    }
     try {
       f.createNewFile();
-      out = new FileOutputStream(f);
+      out = new FileOutputStream(f, false);
 
       outputWriter = new OutputStreamWriter(out);
       outputWriter.write(data);

--- a/app/src/main/java/com/amaze/filemanager/filesystem/ssh/CustomSshJConfig.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/ssh/CustomSshJConfig.java
@@ -23,10 +23,6 @@ package com.amaze.filemanager.filesystem.ssh;
 import java.security.Security;
 
 import net.schmizz.sshj.DefaultConfig;
-import net.schmizz.sshj.signature.SignatureDSA;
-import net.schmizz.sshj.signature.SignatureRSA;
-import net.schmizz.sshj.transport.random.JCERandom;
-import net.schmizz.sshj.transport.random.SingletonRandomFactory;
 
 /**
  * sshj {@link net.schmizz.sshj.Config} for our own use.
@@ -43,15 +39,5 @@ public class CustomSshJConfig extends DefaultConfig {
   public static void init() {
     Security.removeProvider("BC");
     Security.insertProviderAt(new org.bouncycastle.jce.provider.BouncyCastleProvider(), 0);
-  }
-
-  // don't add ECDSA
-  protected void initSignatureFactories() {
-    setSignatureFactories(new SignatureRSA.Factory(), new SignatureDSA.Factory());
-  }
-
-  @Override
-  protected void initRandomFactory(boolean ignored) {
-    setRandomFactory(new SingletonRandomFactory(new JCERandom.Factory()));
   }
 }

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -591,7 +591,7 @@ public class MainActivity extends PermissionsActivity
                         + "\nURL: "
                         + extras.getString(Intent.EXTRA_TEXT);
                 String fileName = Long.toString(System.currentTimeMillis());
-                FileUtil.mktextfile(data, getCurrentMainFragment().getCurrentPath(), fileName);
+                AppConfig.runInBackground(() -> FileUtil.mktextfile(data, getCurrentMainFragment().getCurrentPath(), fileName));
 
                 Toast.makeText(
                         MainActivity.this,

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -591,7 +591,10 @@ public class MainActivity extends PermissionsActivity
                         + "\nURL: "
                         + extras.getString(Intent.EXTRA_TEXT);
                 String fileName = Long.toString(System.currentTimeMillis());
-                AppConfig.runInBackground(() -> FileUtil.mktextfile(data, getCurrentMainFragment().getCurrentPath(), fileName));
+                AppConfig.runInBackground(
+                    () ->
+                        FileUtil.mktextfile(
+                            data, getCurrentMainFragment().getCurrentPath(), fileName));
 
                 Toast.makeText(
                         MainActivity.this,

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -529,9 +529,9 @@ public class MainActivity extends PermissionsActivity
           uris.add(uri);
           initFabToSave(uris);
         }
-          // disable screen rotation just for convenience purpose
-          // TODO: Support screen rotation when saving a file
-          Utils.disableScreenRotation(this);
+        // disable screen rotation just for convenience purpose
+        // TODO: Support screen rotation when saving a file
+        Utils.disableScreenRotation(this);
 
       } else if (actionIntent.equals(Intent.ACTION_SEND_MULTIPLE) && type != null) {
         // save multiple files to filesystem

--- a/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/activities/MainActivity.java
@@ -519,16 +519,20 @@ public class MainActivity extends PermissionsActivity
           zippath = Utils.sanitizeInput(uri.toString());
         }
 
-      } else if (actionIntent.equals(Intent.ACTION_SEND) && type != null) {
-        // save a single file to filesystem
-        Uri uri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
-        ArrayList<Uri> uris = new ArrayList<>();
-        uris.add(uri);
-        initFabToSave(uris);
+      } else if (actionIntent.equals(Intent.ACTION_SEND)) {
+        if (type.equals("text/plain")) {
+          initFabToSave(null);
+        } else {
+          // save a single file to filesystem
+          Uri uri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
+          ArrayList<Uri> uris = new ArrayList<>();
+          uris.add(uri);
+          initFabToSave(uris);
+        }
+          // disable screen rotation just for convenience purpose
+          // TODO: Support screen rotation when saving a file
+          Utils.disableScreenRotation(this);
 
-        // disable screen rotation just for convenience purpose
-        // TODO: Support screen rotation when saving a file
-        Utils.disableScreenRotation(this);
       } else if (actionIntent.equals(Intent.ACTION_SEND_MULTIPLE) && type != null) {
         // save multiple files to filesystem
 
@@ -580,6 +584,23 @@ public class MainActivity extends PermissionsActivity
                       .show();
                   finish();
                 }
+              } else {
+                Bundle extras = intent.getExtras();
+                String data =
+                    extras.getString(Intent.EXTRA_SUBJECT)
+                        + "\nURL: "
+                        + extras.getString(Intent.EXTRA_TEXT);
+                String fileName = Long.toString(System.currentTimeMillis());
+                FileUtil.mktextfile(data, getCurrentMainFragment().getCurrentPath(), fileName);
+
+                Toast.makeText(
+                        MainActivity.this,
+                        getResources().getString(R.string.saving)
+                            + " to "
+                            + getCurrentMainFragment().getCurrentPath(),
+                        Toast.LENGTH_LONG)
+                    .show();
+                finish();
               }
             });
     // Ensure the FAB menu is visible
@@ -2053,7 +2074,9 @@ public class MainActivity extends PermissionsActivity
   }
 
   @Override
-  public void onLoaderReset(Loader<Cursor> loader) {}
+  public void onLoaderReset(Loader<Cursor> loader) {
+    // For passing code check
+  }
 
   private static final class FabActionListener implements SpeedDialView.OnActionSelectedListener {
 

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/SmbConnectDialog.java
@@ -207,7 +207,6 @@ public class SmbConnectDialog extends DialogFragment {
       String userp = "", passp = "", ipp = "", domainp = "";
       conName.setText(name);
       try {
-        jcifs.Config.registerSmbURLHandler();
         URL a = new URL(path);
         String userinfo = a.getUserInfo();
         if (userinfo != null) {

